glslc/src/dependency_info.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015 The Shaderc Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GLSLC_DEPENDENCY_INFO_H
+#define GLSLC_DEPENDENCY_INFO_H
+
+#include <set>
+#include <string>
+
+namespace glslc {
+class DependencyInfoDumpingHandler {
+ public:
+  // Sets -MT argument.
+  void SetDepTargetFileLabel(const std::string& target_label) {
+    dep_target_label_ = std::string(target_label);
+  }
+
+  // Sets -MF argument.
+  void SetDepFileName(const std::string& dep_file_name) {
+    dep_file_name_ = dep_file_name;
+  };
+
+  // Collects the dependency info and dump it to dependency info file.
+  bool DumpDependencyInfo(std::string target_file, std::string source_file_name,
+                          const std::set<std::string>& trace);
+
+  // -MM or -M will prevent emitting the output of compilation.
+  bool dump_dependency_info_only_ = false;
```

Comment:
Hmm, I was thinking about this, but there is a corner case: when both -M and -MD are specified.

In Clang++, it makes -M overrides -MD and emits an warning to say -MD is not used.
In GCC, it generates dependency info files with name: *.d, and blocks compilation output.

In either way, we may need another flag, as we actually have three states, instead of two.
Clang: we need to know if both -MD and -M are specified, so we need one flag for each of them.
GCC: As you said, -M writes dependency info to main output and -MD writes out separately. But this means, in '-M -MD' cases, if we have -MD overrides -M, we will have compilation output, but if we let -M overrides -MD, we will have dependency info dumped in a wrong stream.

In general I think there are three states instead of two. I was wondering how do we want glslc to behave in such cases.